### ISSUE TITLE
Memory Leak when using top-level await before defining computed translations

### ISSUE THREAD

**Issue by azhirov** [nuxt-modules/i18n#2629]
### Environment

- Operating System: Linux
- Node Version:     v20.8.1
- Nuxt Version:     3.8.2
- CLI Version:      3.10.0
- Nitro Version:    2.8.1
- Package Manager:  npm@10.1.0
- Builder:          -
- User Config:      app, modules, devtools
- Runtime Modules:  @pinia/nuxt@0.5.1, @nuxtjs/i18n@8.0.0-rc.11
- Build Modules:    -

### Reproduction

I've created a minimal reproduction [repository](https://github.com/azhirov/nuxt-i18n-pinia-memory-leak) to illustrate the problem. To reproduce the issue:

1. Clone the repository: `git clone git@github.com:azhirov/nuxt-i18n-pinia-memory-leak.git`
2. Install dependencies: `npm install`
3. Build the project: `npm run build`
4. Run the server: `node .output/server/index.mjs`
5. Use a tool like `ab`, `siege`, or any other utility to send multiple requests. For example: `siege -c10 -t30s http://127.0.0.1:3000` or `ab -n 5000 -c 100 http://localhost:3000/`
6. Check the memory consumption of the process (for example, using the `top` command).

### Describe the bug

I've encountered a peculiar memory leak issue related to the simultaneous usage of the `@nuxtjs/i18n` and `@pinia/nuxt` modules in a Nuxt 3 project. The problem manifests when an asynchronous method from the Pinia store is called (using `await`) before a `computed` property that includes `useNuxtApp().$i18n.t()`. This results in a memory leak. Interestingly, moving the asynchronous method below the `computed` property prevents the leak. Additionally, if the `$i18n.t()` is not used in the computed property, the leak does not occur.

Code with memory leak:
```
import useFooStore from '~/store/foo';

const fooStore = useFooStore();
await fooStore.loadData()

const text = computed(() => {
  const i18n = useNuxtApp().$i18n;
  return i18n.t('seventeen');
})
```

After running the command `node .output/server/index.mjs`, the server consumes approximately 20 MB of RAM. However, after executing the command `siege -c10 -t30s http://127.0.0.1:3000`, the memory usage increases to 1.4 GB and remains at approximately this level even after stopping the load.

![node_1091683](https://github.com/nuxt-modules/i18n/assets/29061184/00714846-5f98-4612-b37f-af750da0f6b1)

If you move the `await` call to the asynchronous method of the store below the computed property, the graph looks like this (the memory leak is significantly reduced, and memory consumption resets to almost the initial level).

![node_1093319](https://github.com/nuxt-modules/i18n/assets/29061184/6cb4acc1-1690-412e-af45-6e9cdc4d2698)

I've noticed an interesting pattern: the memory leak seems to intensify with an increase in the number of translation files (in the `lang` directory).

### Additional context

A similar issue was mentioned earlier in this comment: https://github.com/nuxt-modules/i18n/issues/2034#issuecomment-1659386170. After updating to version `8.0.0-rc.11`, the problem reproduces only when using Pinia. A simple asynchronous function defined within the same component does not trigger the memory leak.

**Comment by BobbieGoede** [comment]
Thanks for reporting and providing a reproduction! 

This sounds similar to https://github.com/nuxt-modules/i18n/issues/2034, I thought I fixed that 😥 hopefully I'll have time to debug this soon.

**Comment by agracia-foticos** [comment]
Same issue, i have stores before computed properties with t()

**Comment by BobbieGoede** [comment]
Hmm while working on fixing the other memory leak in this PR https://github.com/nuxt-modules/i18n/pull/2616 I did also test it on this reproduction https://github.com/danielroe/i18n-memory-leak and at that time it did seem to work.

@kazupon @danielroe 
Maybe my testing method is flawed, if I use `pnpm link [path to local i18n module]` and build and test https://github.com/danielroe/i18n-memory-leak, the memory leak seems to be gone or minimal, but if I install the edge channel which should be essentially the same, the memory leak is present.

It's clear I need to read up on exactly what `pnpm link` does, but could this hint to the issue being related to a dependency (seeing as the lockfiles differ)? The memory leak that I did find and fix with https://github.com/nuxt-modules/i18n/pull/2616 did not behave differently when using `pnpm link`.

**Comment by azhirov** [comment]
@BobbieGoede I tried your method with my [repo](https://github.com/azhirov/nuxt-i18n-pinia-memory-leak). I cloned `@nuxtjs/i18n`, installed dependencies (`pnpm install`), built it (`pnpm run build`), and then installed it locally in my reproduction repository. 

However, the memory leak still persists, both in the `main` branch and on the `v8.0.0-rc.11` tag.

**Comment by BobbieGoede** [comment]
@azhirov 
Oh I failed to mention that I tested that as well and this wasn't the case with your reproduction, can you confirm that linking it locally in https://github.com/danielroe/i18n-memory-leak does seem to (mostly) resolve the memory leak? Trying to figure out what makes the difference but I maybe looking in the wrong direction 😅..

**Comment by azhirov** [comment]
@BobbieGoede 
Sorry, I can't confirm that. I tested it with https://github.com/danielroe/i18n-memory-leak, and the memory leak persists even with the locally installed `@nuxtjs/i18n` module. This is true for both the `main` branch and the `v8.0.0-rc.11` tag. Perhaps I am doing something wrong :(

**Comment by kazupon** [comment]
@BobbieGoede 

> Maybe my testing method is flawed, if I use `pnpm link [path to local i18n module]` and build and test https://github.com/danielroe/i18n-memory-leak, the memory leak seems to be gone or minimal, but if I install the edge channel which should be essentially the same, the memory leak is present.

I often build nuxt i18n as gzip with `pnpm pack` and add it to the target repro project with file protocol (`file://`) in deps.
This way has the advantage that the package manager is not affected by the link and implicitly by the package manager, and is completely isolate.

**Comment by BobbieGoede** [comment]
@kazupon 
Ah I have used that too, I was hoping `pnpm link` would work the same without needing to zip and install 😥, my workflow for checking the memory leak is a bit slow.

@azhirov 
No worries, I don't think the difference between the two reproductions matter much after all.

What's interesting is that this memory leak doesn't seem to be present if the translations are in `i18n.config.ts` instead of loading them in files. Still working on debugging this!

**Comment by BobbieGoede** [comment]
For some reason the memory leak is also gone when moving the computed translation inside a component. I forked your reproduction and added that workaround if you would like to try: https://github.com/BobbieGoede/nuxt-i18n-2629-memory-leak.

page (`pages/no-leak-component.vue`)
```ts
<script setup lang="ts">
import useFooStore from "~/store/foo";
import TranslationComponent from "~/components/TranslationComponent.vue";

const fooStore = useFooStore();
await fooStore.loadData();
</script>

<template>
	<div>
		<TranslationComponent />
	</div>
</template>
```

component (`components/TranslationComponent.vue`)
```ts
<script setup lang="ts">
const text = computed(() => {
	const i18n = useNuxtApp().$i18n;
	return i18n.t("seventeen");
});
</script>

<template>
	<div>
		{{ JSON.parse(JSON.stringify(text)) }}
	</div>
</template>
```

Perhaps this could work as a workaround, besides the known workaround of moving await after the computed variables. Will continue searching for the source of the issue but lower the priority label due to the issue's specificity.

**Comment by azhirov** [comment]
@BobbieGoede Certainly, this is an interesting approach, and it does address some issues.

However, I disagree that it can be a full-fledged workaround, as the result of `computed` might be used, for example, on the page - to be passed to `useHead`. In this case, the value may also depend on `useAsyncData`. As far as I understand, it's not possible to move this to a separate component, as the SSR result will differ from rendering on the client side. 

I believe that the bug's priority remains "important", considering the widespread use of this module, including in production.

**Comment by BobbieGoede** [comment]
I'm not sure I fully understand your use case, as I understand it you can use derived values in `computed` regardless of whether these are fetched further down the script. 

I have changed my fork to demonstrate this shouldn't cause SSR issues and changed the Pinia store to fetch from a server route, see https://github.com/BobbieGoede/nuxt-i18n-2629-memory-leak/blob/master/pages/index.vue for example.

I lowered the label priority as I think the usage is specific enough that I feel this shouldn't block v8 release, preventing users from migrating from v7 to v8. I'm not sure how widespread the usage is, if more projects are running into this I would like to know, I'll raise the priority again either way, as it does impact performance as the label describes.

**Comment by szwenni** [comment]
Hi
I'm also impacted by this problem.
The problem is we have a very very big project where this occured when we switched from loading from json files to a loader file and then having a async function which returns the tranlsations from a server response.

Two things I observed, messages loaded from a function are not cached at all, maybe this should be somehow configurable. If the lazy loaded messages are static why not cache them ?

Second thing, following the "Allocation instrumentation on timeline" with the chrome inspector I observed that the leak object is coming from here: https://github.com/nuxt-modules/i18n/blob/4293f311500339d25b9eeb1e69aa4de10cb7f003/src/runtime/messages.ts#L59
I can till now not really tell why the objects are still referenced from somewhere, but each reload of the page a new object with all messages for a locale is spawned on the heap and keeps being referenced so GC is not triggered.

For us in production this leads to a crash of the application every x requests after the heap is filled.
As mentioned the application is quite big so I'm not really able to identify right now why this things are still referenced in some map somewhere and which WeakMap exactly is still referencing it. It looks like either the ssr payload or the state but I'm not able to verify it.

What I can surely tell is that we have no async pinia function which is executed on server side as all of our async pinia stuff is client side only (onMounted)

**Comment by s00d** [comment]
It seems I managed to localize and fix the issue.

https://github.com/s00d/i18n/tree/fix-leak

In the Nuxt documentation, they recommend using JSON instead of a class. I rewrote your code to import JSON, and the problem disappeared. Therefore, the memory leak is definitely related to generateTemplateNuxtI18nOptions.

Here is the documentation link.

https://nuxt.com/docs/api/kit/templates#examples

Here are my changes.

https://github.com/s00d/i18n/commit/c6e4163d2ca7c00337eb34bfb9539a2de349c606
https://github.com/s00d/i18n/commit/d05a972a12925d2a3bca217312541afc7a432b8b

Please note that this is just a code example, it's not very functional and triggers warnings due to JSON imports, but I believe it will help in identifying the issue.

**Comment by s00d** [comment]
It seems the problem is here.

https://github.com/nuxt-modules/i18n/blob/b9c250e8ff64a720145c4389a862b6977adc806f/src/gen.ts#L74

If we refactor it like this

```ts
importer.load = JSON.stringify(genDynamicImport(importSpecifier, !isServer ? { comment: `webpackChunkName: "${meta.key}"`, assert: { type: 'json' } } : { assert: { type: 'json' } }))

//...

const loadFunction = new Function('return ' + load)();
const getter = await loadFunction().then(r => r.default || r)

```


and also fix a few other places, for example, the path to the locale file, then the memory leak disappears.

In my example above, there are some changes that need to be made, but when switching to JSON, a lot of other problems arise. We need to find a different solution.

**Comment by szwenni** [comment]
After testing out a lot of stuff and trying to dig down the bug, I actually found out that for me it was related to the nuxt experimental feature called asyncContext.

When running with async context I can produce relative reliable a OOM with the leaking object (the loaded translations), when deactivated it just works fine.

For reference the description of the asyncContext feature
https://github.com/nuxt/nuxt/pull/20918

**Comment by szwenni** [comment]
I made a reproduction here
https://github.com/szwenni/nuxt-i18n-memory-leak

I think one workaround, which is most probably not that clean, would be, when the messages are loaded for the first time for a language, store them in a global cache and then on the next loads just reference them from there instead of deepCloning them every time on every request. I'm not 100% sure what the intention behind the deepCloning is/was.

I would volunteer to do a pull request if nothing speaks against leaving out the deepCloning every time but just do it once on initial load for a locale. This would also save some GC time as it will result in a lot less allocations especially when you have big translation files.

Maybe it makes also sense to keep the deepCloning as default behavior but add an option which allows this caching to be enabled and circumvent the deepCloning of all messages every time.

**Comment by agracia-foticos** [comment]
Versión 8 final estable does memory leak also
![image](https://github.com/nuxt-modules/i18n/assets/77386602/b2017e15-e52d-40a8-8fbe-bed94e647449)

**Comment by BobbieGoede** [comment]
@s00d 
I'm unable to run this locally with the playground, I think the leak is not present because it may not be loading all translations. I think your changes would break dynamic translations as well but I haven't checked further yet.

@agracia-foticos 
Unfortunately we haven't fixed this issue yet so it isn't included in the stable release, I'm still working hard on finding a fix.

@szwenni 
Deep cloning happens because it's possible for locale files to fetch translations from a remote source or differ based on runtime config, I think it's possible to reduce the amount of deep cloning that happens but the order/priority of locale files would have to be taken into account as well as they can overwrite each other.

**Comment by szwenni** [comment]
For initial creating and loading the locale messages this seems okay but the deep cloning is happening on every route navigation right now (I checked it with the debugger).

I will try to implement my suggestion and you can check it if it makes sense or not.

In my repro I‘m using remote fetching with a loader. I did not check with local files which don‘t go through a loader and a remote api.

**Comment by BobbieGoede** [comment]
@szwenni 
For me the confusion is that the leak isn't present when the computed translation happens before an awaited function or if there is no awaited function at all.

I think reducing deep cloning and loading where possible would probably improve performance and memory usage, but since these things happen in the setups without memory leak as well I think it likely won't fix the leak itself.

**Comment by s00d** [comment]
> @s00d I'm unable to run this locally with the playground, I think the leak is not present because it may not be loading all translations. I think your changes would break dynamic translations as well but I haven't checked further yet.

There, half of everything doesn't work at all. I only debugged the server side; I broke everything else. But the point is, if you change importer.load, the leak disappears, and I see the logic in that. Since file loading always happens asynchronously, there might be some kind of conflict.

I would carefully look at this code:

```ts
importer.load = genDynamicImport(importSpecifier, !isServer ? { comment: `webpackChunkName: "${meta.key}"` } : {}) 
} else {
importer.load = () => Promise.resolve(${meta.key})
}
```

The leak is most likely occurring because of this piece of code

**Comment by BobbieGoede** [comment]
@s00d 
Then the question remains, why doesn't this code leak if no (Pinia) top-level await is used?

I'm working on another reproduction that should further demonstrate when this leak happens and when not, after more testing it may be related to Pinia after all.

**Comment by s00d** [comment]
> @s00d Then the question remains, why doesn't this code leak if no (Pinia) top-level await is used?
> 
> I'm working on another reproduction that should further demonstrate when this leak happens and when not, after more testing it may be related to Pinia after all.

It doesn't seem like Pinia; I also have a leak [here](https://github.com/danielroe/i18n-memory-leak). I tried installing from the repo and from npm, and it's there everywhere

**Comment by szwenni** [comment]
> @szwenni For me the confusion is that the leak isn't present when the computed translation happens before an awaited function or if there is no awaited function at all.
> 
> I think reducing deep cloning and loading where possible would probably improve performance and memory usage, but since these things happen in the setups without memory leak as well I think it likely won't fix the leak itself.

The things is my reproduction ist neither using computed props nor awaits. Just very basic two pages and the leak is still there.

My assumption that the fixing the deppCloning with additional caching should fix the memory leak is based on the fact that the leaking objects are created by th „setter“ function in the initialLoadMessages. (The one I linked above). If you always return the same reference to a array here it wont leak as it will not be produced again (so much to the theory)

**Comment by BobbieGoede** [comment]
@s00d 
I've been changing and testing reproductions a lot locally especially when the reproduction is older, the leak is still present but reduced in https://github.com/danielroe/i18n-memory-leak when making the following change:

```diff
-const { $i18n } = useNuxtApp();
+const { t } = useI18n();

await new Promise((resolve) => setTimeout(resolve, 10));

-const timeLabel = computed(() => $i18n.t("general"));
+const timeLabel = computed(() => t("general"));
```

In my testing the leak is more severe specifically when async Pinia actions are used top-level. But you're right the leak is present in varying degrees as long as an async function is used before defining a computed variable with a translation.

Either way, I have my doubts that this is related to the loading of translations, because this should be pretty much the same regardless of what happens in the page or component.

@szwenni 
Apologies, I somehow missed your reproduction entirely. I will need to take a good look at it later, it's possible that you are running into a different leak entirely.


As for the leak in this issue, I suspect the `t` function is tracking the context or at least the loaded translations for reactivity and using async functions before it somehow breaks this and prevents it from being cleaned up.

**Comment by szwenni** [comment]
The more I read it here the more I think I‘m trigerring a different leak and additionally also the same leak in production when having many requests. Kind of a false positive for this bug.

initially it seemed like the same leak but I think these are two different leaks.

Did you figure out already here which data structure gets leaked over time ?

**Comment by s00d** [comment]
@szwenni @BobbieGoede
I think you are mistakenly assuming these are different issues. For me, the leak occurs exactly when calling

```js
load().then(...
```

After I changed the translation retrieval method, the leak immediately stopped. It's possible that the leak's origin varies, but it seems very likely that the main problem is here.

It's possible that the leak diminishes when using useI18n because there are fewer calls to the module, meaning the leak just reduces.

**Comment by szwenni** [comment]
I also have the feeling both issues relate to the same base bug.

Did you figure out already which object gets leaked on your side ? For me it has been the (at least) translations which had been created by the deepCopy in the setter function in the loadLocale function.

I‘m still trying to figure out the link between the leak and your proposed fixed with using json. I would really like to understand it.

**Comment by s00d** [comment]
@szwenni 
I didn't understand the reason myself. The difference is that in my code, there is no direct import from the config, and most likely because of this, a memory leak is not occurring somewhere further. I've tried a few other possible places, but so far, I haven't been able to catch anything.

**Comment by szwenni** [comment]
okay
So switching off the asyncContext feature for me actually solved the "big" memory leak where the translations had been leaked.
But still some memory leak is happening somewhere (slower than before). Reloading the page a few times especially in dev mode is still filling up the heap a lot.

I need to investigate if I can find the object which is getting leaked but most probably some configuration object of the nuxt internals.

**Comment by s00d** [comment]
https://github.com/nuxt-modules/i18n/pull/2647

Basically, it seems that I found where the leak is. If you do as in my pull request, the leak disappears. It means that a reference is created in 'messages' and can't be destroyed. I spent a lot of time searching, but after the fix, everything started working fine. Unfortunately, due to such a solution, the framework began to work slower, but I think you will figure it out from here.

**Comment by s00d** [comment]
I've added new information [here](https://github.com/nuxt-modules/i18n/pull/2647#issuecomment-1876557005), and it's all much simpler than I thought. The issue lies with deepCopy, and my fix worked because the references in messages were removed.

**Comment by BobbieGoede** [comment]
@s00d 
Can you check if #2654 works for you locally? It combines your findings about `deepCopy` with some optimizations.

Also, do you also still experience a small memory leak with the fixes so far? I suspect the messages leaking is a symptom of the context or i18n options leaking somewhere later in the code, but since the messages take up the largest amount of memory/space of the options it ends up being the most impactful.

**Comment by s00d** [comment]
> @s00d Can you check if #2654 works for you locally? It combines your findings about `deepCopy` with some optimizations.
> 
> Also, do you also still experience a small memory leak with the fixes so far? I suspect the messages leaking is a symptom of the context or i18n options leaking somewhere later in the code, but since the messages take up the largest amount of memory/space of the options it ends up being the most impactful.


Sorry, I didn't check which pull request you were writing about, I'll check now.

Locale retrieval is not working.|

Regarding structuredClone, I tested it a couple of weeks ago; it operates slower than JSON.parse(JSON.stringify()).

**Comment by BobbieGoede** [comment]
Can you check if you're still experiencing the leak using the latest edge release? (`npm i -D @nuxtjs/i18n@npm:@nuxtjs/i18n-edge@latest`)

I made some changes in `vue-i18n` (https://github.com/intlify/vue-i18n-next/pull/1677) that should reduce memory usage and it seems the messages are not leaking (as much) anymore. I still don't think this removes the leak fully, a very small leak is likely still present.

**Comment by wizzout** [comment]
> Can you check if you're still experiencing the leak using the latest edge release?

I've checked on my own stand with same problem - no more memory leaks! Total heap size increases when there ara many active request, but then return to normal value fast. Really waiting release of this fix! Thank You!

**Comment by BobbieGoede** [comment]
@wizzout 
Thanks for checking the fix! Hopefully we can release this fix soon, awaiting more confirmation, it seems like the leak may still be present for some setups (https://github.com/nuxt-modules/i18n/issues/2612#issuecomment-1884558040)

@azhirov @s00d 
Can you confirm if the leak has been resolved in your project using the latest edge release? You may need to remove `node_modules` and lockfiles before installing.

**Comment by s00d** [comment]
> @azhirov @s00d
> Can you confirm if the leak has been resolved in your project using the latest edge release? You may need to remove `node_modules` and lockfiles before installing.

Checked on one of the test repositories, everything is fine now. Unfortunately, it will take significantly more time for the others. For now, we have worked around the issue with temporary solutions. I will try to do more testing.

**Comment by azhirov** [comment]
I have checked several versions of the module for memory leaks. The repository https://github.com/azhirov/nuxt-i18n-pinia-memory-leak was used for testing. Before installing a new version each time, I deleted `package-lock.json` and the `node_modules` directory. Then I built the project and ran it: `nuxi build && node .output/server/index.mjs`

For load testing, the command `siege -c10 -t30s http://127.0.0.1:3000` was used.

Here are the results:

### 8.0.0-rc.5 ✅
No leaks, no errors in console output (`"vue-i18n": "9.5.0"`)

<details>
  <summary>RAM usage graph (click to view)</summary>

  ![rc5](https://github.com/nuxt-modules/i18n/assets/29061184/e0b981ca-f3f2-4002-9e9d-62dd217361cd)
</details>

### 8.0.0-rc.6 ❌
There is a leak, and there is an error `[nuxt] [request error] [unhandled] [500] Maximum call stack size exceeded`

<details>
  <summary>RAM usage graph (click to view)</summary>

  ![rc6](https://github.com/nuxt-modules/i18n/assets/29061184/ef558e21-f82a-439e-b9d2-e30f09af457b)
</details>

### 8.0.0-rc.7 ❌
There is a leak, and there is an error `[nuxt] [request error] [unhandled] [500] Maximum call stack size exceeded`

<details>
  <summary>RAM usage graph (click to view)</summary>

  ![rc7](https://github.com/nuxt-modules/i18n/assets/29061184/a46be87e-5fae-4b7e-9c94-e1c576da5875)
</details>

### 8.0.0-rc.8 ❌
There is a leak, and there is an error `[nuxt] [request error] [unhandled] [500] Maximum call stack size exceeded`

<details>
  <summary>RAM usage graph (click to view)</summary>

  ![rc8](https://github.com/nuxt-modules/i18n/assets/29061184/3b8d6e17-3bc6-4084-9ad2-21130f3b8b4e)
</details>

### 8.0.0-rc.9 ❌
There is a leak, and there is an error `[nuxt] [request error] [unhandled] [500] Maximum call stack size exceeded`

<details>
  <summary>RAM usage graph (click to view)</summary>

  ![rc9](https://github.com/nuxt-modules/i18n/assets/29061184/b770bc29-9008-4ae9-a10b-b56637e7d4a8)
</details>

### 8.0.0-rc.10 ❗️
No leaks, but there is an error `[nuxt] [request error] [unhandled] [500] Maximum call stack size exceeded`

<details>
  <summary>RAM usage graph (click to view)</summary>

  ![rc10](https://github.com/nuxt-modules/i18n/assets/29061184/92b84271-ee71-4156-9efe-b0ac493e364f)
</details>

### 8.0.0-rc.11 ✅
No leaks. No errors. I was surprised here because this version had the issue. I examined the `package-lock.json` file and found that a new version of the `vue-i18n` package (`9.9.1`) is now installed.

<details>
  <summary>RAM usage graph (click to view)</summary>

  ![rc11-vue-i18n-9 9 1](https://github.com/nuxt-modules/i18n/assets/29061184/c2052d87-4234-49c9-b3c2-7f91dd29b80a)
</details>

### 8.0.0-rc.11 and vue-i18n@9.8.0 ❌
There is a leak. To check, I deliberately installed an old version of `vue-i18n` using the command `npm i --save vue-i18n@9.8.0`. No errors in console output.

<details>
  <summary>RAM usage graph (click to view)</summary>

  ![rc11-vue-i18n-9 8 0](https://github.com/nuxt-modules/i18n/assets/29061184/f9750993-b2f8-47bb-a84b-cd4451350b80)
</details>

### i18n-edge@8.0.0-28444362.1cb9e70 ✅
No leaks. Probably because the minimum version of the package is specified in the dependencies: `"vue-i18n": "^9.9.0"`. No errors in console output.

<details>
  <summary>RAM usage graph (click to view)</summary>

  ![edge](https://github.com/nuxt-modules/i18n/assets/29061184/bf6271dd-390f-43c9-b7df-14684e653532)
</details>

---

I might be mistaken, but it seems that the leak was in the `vue-i18n` package, and it is fixed in version `9.9.0`. Therefore, in the production environment, it is sufficient to specify in `package.json`:

```json
"@nuxtjs/i18n": "8.0.0",
"vue-i18n": "^9.9.0"
```

No need to install `@nuxtjs/i18n-edge@latest`.

**Comment by BobbieGoede** [comment]
@azhirov 
Thank you for your thorough testing, this is great!

You're right, most if not all of the memory leak was fixed in `vue-i18n@9.9.0`, a fresh install of `@nuxtjs/i18n@^8.0.0` should use this version as well (as you noticed with `rc.11`). There have been a few changes since release so it's good to see that the edge release is still doing well memory wise in your setup.

Will soon publish a patch release with the recommendation to refresh lockfiles when updating to ensure the latest `vue-i18n` is installed along with it. After which I'll close this issue, thanks again for providing your minimal reproduction and this detailed report!

**Comment by szwenni** [comment]
Hello all
for me the problem still exists. I can see in the heap dump of the nuxt process a high amount of copies of the messages for each language, all referenced in different vueI18nOptions objects.

For context we have around 40 different languages, and after some time (around 20 mins) it has filled up all the available heap, crashes and restarts. Only then the heap is again freed but till then it always fills up everything.

A heap dump created after 10 minutes in our productive environment shows around 50 copies of the 'de' messages object (which is the default language) in 50 different vueI18nOptions objects. And this piles up.

So somehow this problem is still not fixed. From configuration point of view I changed nothing than the things already described above.

I'm not sure how to make a reproduction if this entirely and if I should create a new issue or this one can be reopened.

**Comment by BobbieGoede** [comment]
@szwenni 
Since we're unable to trigger a memory leak with the reproductions in the closed issues you're likely experiencing a different issue/leak. Please open a new issue with a reproduction so we can debug it.

**Comment by szwenni** [comment]
@BobbieGoede 
I just tested it with my repdroduction and it stills triggers it.
https://github.com/szwenni/nuxt-i18n-memory-leak
If you run this locally, take a heap snap-shot at the beginning, cycle through some languages, and take a snapshot again.
In the array section of the heap snapshot you will see a lot of objects with the same size at end and many objects are references to a messages object from a vueI18nOptions object. And like said these are often from the same language.

**Comment by liorchamla** [comment]
> I might be mistaken, but it seems that the leak was in the `vue-i18n` package, and it is fixed in version `9.9.0`. Therefore, in the production environment, it is sufficient to specify in `package.json`:
> 
> ```json
> "@nuxtjs/i18n": "8.0.0",
> "vue-i18n": "^9.9.0"
> ```
> 
> No need to install `@nuxtjs/i18n-edge@latest`.

Hello to all, we had a problem with a newly push to production Nuxt 3 app. This modification to packages.json made the trick in local. We will deploy a new release of our app soon in prod container. 

For context : 
- Node 18.16.0 with old dependencies + Artillery test scenarios : the app goes up to 2GB RAM and does not come back even after the tests end
- Node 18.16.0 with new dependencies + Artillery test scenarios : the app reaches 800MB MAX and the RAM itself is sooo much variable, even when the tests are running, we see the garbage collector at work.

So, to conclude, @azhirov : thank you so much mate.

**Comment by vhovorun** [comment]
@szwenni Hey have u found the solution to your leak? Seems like I have the same problem

**Comment by BobbieGoede** [comment]
@vhovorun 
Can you open a new issue with a minimal reproduction? 🙏

**Comment by szwenni** [comment]
So actually there are two sides I think
* I have the feeling there is still a memory leak somewhere but it is quite hard to identify which object leaks
* second thing is that the createMatcher / match route method is quite slow when having a lot of routes

This both things together and additionally node being under pressure (a lot of requests, e.g. google is traversing your site) leads to node no having the time to gc fast enough also the big objects and then finally lead to a oom

I for myself fixed this by using someones pull request where he adopted the routematcher to not use only regex but also a tree. This actually speed up the whole application by 1000% and till now I did not have any issues.
https://github.com/vuejs/router/pull/2148

**Comment by BobbieGoede** [comment]
@szwenni 
Yep the amount of routes has a significant impact on performance since Vue Router v4 (which Nuxt 2 does not use), we're tracking that here https://github.com/vuejs/router/issues/2132, this is more noticeable for i18n projects due to the generated routes. Have you also checked the performance using Vue Router `4.3.3` or higher? This includes https://github.com/vuejs/router/pull/2137 and should be much more performant.

Of course if https://github.com/vuejs/router/pull/2148 works well that's great, there's also room for the i18n module to improve routing performance (and build size) by changing the matching approach such as https://github.com/nuxt-modules/i18n/pull/2937. It's all an ongoing effort 😅

**Comment by s00d** [comment]
> @szwenni Yep the amount of routes has a significant impact on performance since Vue Router v4 (which Nuxt 2 does not use), we're tracking that here [vuejs/router#2132](https://github.com/vuejs/router/issues/2132), this is more noticeable for i18n projects due to the generated routes. Have you also checked the performance using Vue Router `4.3.3` or higher? This includes [vuejs/router#2137](https://github.com/vuejs/router/pull/2137) and should be much more performant.
> 
> Of course if [vuejs/router#2148](https://github.com/vuejs/router/pull/2148) works well that's great, there's also room for the i18n module to improve routing performance (and build size) by changing the matching approach such as #2937. It's all an ongoing effort 😅

Hello. After careful consideration, I've decided to preserve all my work on this project. I've invested a significant amount of time and effort into it, and I'm quite proud of the results.

https://github.com/s00d/nuxt-i18n-micro

Perhaps others have encountered similar challenges. I've detailed the issues and solutions in the README. Currently, my module offers a 2-3x performance improvement in most cases, and up to a 10x speedup in some scenarios. It's also smaller, simpler, and easier to understand.

**Comment by BobbieGoede** [comment]
@s00d 
The performance comparison looks impressive! Are you planning on supporting a specific set of features or working towards a similar feature set as this module?

Having more modules explore i18n should make for a healthier ecosystem 💪

**Comment by s00d** [comment]
> The performance comparison looks impressive! Are you planning on supporting a specific set of features or working towards a similar feature set as this module?
> 
> Having more modules explore i18n should make for a healthier ecosystem 💪


It's difficult to say what the future holds right now. My current focus is on understanding how Vite and Nuxt work together, and I'm not just optimizing the build process, but also enhancing the module's runtime performance. I chose not to use Vue i18n due to the numerous issues I've personally encountered, which has brought both advantages and challenges. At the moment, my priority is catching bugs and getting the module ready for production. After that, we'll see what direction things take.

I'm also exploring how to properly implement type definitions for the Nuxt plugin. So far, I haven't managed to create a fully functional solution—I've only been able to achieve some level of typing in the IDE through composables. Unfortunately, neither the documentation nor GitHub searches have provided clear answers. There are still a lot of hurdles to overcome before I can start thinking about expanding the module's features.

**Comment by vhovorun** [comment]
@BobbieGoede Hey, I found the problem and it wasn't obvious at all 😅 but maybe it will save for someone a lot of time, in my case the fix was switching from 22.5.1 node version to 22.0.0 after that memory goes high but it resets back to normal
